lib/client-assets.php

```diff
@@ -693,3 +693,27 @@ function gutenberg_extend_settings_custom_units( $settings ) {
 	register_block_pattern( 'core/its-time', gutenberg_load_block_pattern( 'its-time' ) );
 	register_block_pattern( 'core/testimonials', gutenberg_load_block_pattern( 'testimonials' ) );
 }
+
+/**
+ * Extends block editor settings to include UI labels to be shown for elements.
+ *
+ * Currently experimental.
+ *
+ * @param  array $settings Default editor settings.
+ * @return array           Modified editor settings.
+ */
+function gutenberg_extend_settings_ui_labels( $settings ) {
+	$post_type        = get_post_type();
+	$post_type_object = get_post_type_object( $post_type );
+
+	$document_label = in_array( $post_type, array( 'post', 'page' ), true ) ?
+		__( 'Document', 'gutenberg' ) :
+		$post_type_object->labels->singular_name;
+
+	$settings['__experimentalLabels'] = array(
+		'document' => $document_label,
+	);
+
+	return $settings;
+}
+add_filter( 'block_editor_settings', 'gutenberg_extend_settings_ui_labels' );
```

Comment:
I think this php code should be useless if we rely on the same mechanism (post type) to retrieve the labels used in other components.

Reply:
If we don't provide a filter, how do you imagine other plugins can override this value?

My use case is: I have a post meta post category (possible values are Newsletter, Story, Collection, Article etc.) which is not directly related to the core post type. So I need to be able to filter this label value from somewhere.

Reply:
Curious why you're using a post meta for that and not a separate CPT for each one of these.

Reply:
> Curious why you're using a post meta for that and not a separate CPT for each one of these.

Our CMS is closed source for our editorial workflow. So won't be able to give direct example of our architecture. But in summary, we started with two article type Standard and Feature article. Data, input fields and api design for both is same till date. For the most part, it was to only have an affect on the render layer with styles and page layout. Eventually, we ended up having other types such as Newsletters, Live Blog Posts etc. They differ a bit in terms of certain meta fields. But they inherit from the same structure as Standard or Feature Article. So we chose to have a post meta for this. Today, for us, it's mostly about supporting that historical decision in the existing system.

Reply:
I'm happy to keep the same source of truth for the post type labels, as long as we have a place to filter it out.

As you said, I can use below as is in SettingsHeader component. But then I may have to add the JS filter which I had added earlier and we moved away from that approach.
```
const { getCurrentPostType } = select( 'core/editor' );
const { getPostType } = select( 'core' );
const postType = getPostType( getCurrentPostType() );
const documentLabel = postType.labels.singular_name
```

Reply:
I understand your use-case but tbh I'm not sure it warrants an extensibility API and the need to maintain it while the CPT label seems like the most logical choice here.

Reply:
@youknowriad I think I will be able to work around this approach which you have suggested. But I will make changes in the PR accordingly. But I'm intrigued by @peterwilsoncc 's suggestion in the comments. It would be worth giving a thought to it.